perun-wui-registrar/src/main/java/cz/metacentrum/perun/wui/registrar/widgets/items/WidgetBox.java

```diff
@@ -0,0 +1,127 @@
+package cz.metacentrum.perun.wui.registrar.widgets.items;
+
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.user.client.ui.HasHorizontalAlignment;
+import com.google.gwt.user.client.ui.HorizontalPanel;
+import com.google.gwt.user.client.ui.VerticalPanel;
+import com.google.gwt.user.client.ui.Widget;
+import cz.metacentrum.perun.wui.json.Events;
+import cz.metacentrum.perun.wui.model.beans.ApplicationFormItemData;
+import cz.metacentrum.perun.wui.registrar.widgets.PerunForm;
+import cz.metacentrum.perun.wui.registrar.widgets.items.validators.PerunFormItemValidator;
+import cz.metacentrum.perun.wui.registrar.widgets.items.validators.WidgetBoxValidator;
+import cz.metacentrum.perun.wui.widgets.PerunButton;
+import org.gwtbootstrap3.client.ui.constants.ButtonType;
+import org.gwtbootstrap3.client.ui.constants.IconType;
+import org.gwtbootstrap3.client.ui.html.Paragraph;
+
+/**
+ * WidgetBox represent shared logic for ListBox and MapBox.
+ *
+ * @author Jakub Hejda <Jakub.Hejda@cesnet.cz>
+ */
+public class WidgetBox extends PerunFormItemEditable {
+
+	private final WidgetBoxValidator validator;
+	protected Widget widget;
+
+	public WidgetBox(PerunForm form, ApplicationFormItemData item, String lang) {
+		super(form, item, lang);
+		this.validator = new WidgetBoxValidator();
+	}
+
+	@Override
+	public String getValue() {
+		return null;
+	}
+
+	@Override
+	public void validate(Events<Boolean> events) {
+		validator.validate(this, events);
+	}
+
+	@Override
+	public boolean validateLocal() {
+		return validator.validateLocal(this);
+	}
+
+	@Override
+	public PerunFormItemValidator.Result getLastValidationResult() {
+		return null;
+	}
+
+	@Override
+	public boolean focus() {
+		return false;
+	}
+
+	public Paragraph getPreview() {
+		if (widget instanceof Paragraph) {
+			return (Paragraph) widget;
+		}
+		return null;
+	}
+
+	@Override
+	protected Widget initWidget() {
+		VerticalPanel vp = new VerticalPanel();
+		vp.setWidth("100%");
+		vp.setHorizontalAlignment(HasHorizontalAlignment.ALIGN_CENTER);
+		PerunButton addButton = generateAddButton(vp);
+		addButton.setIcon(IconType.PLUS);
+		addButton.setType(ButtonType.SUCCESS);
+		addButton.setTooltipText(getTranslation().addNewValue());
+
+		vp.add(addButton);
+		return vp;
+	}
+
+	@Override
+	protected Widget initWidgetOnlyPreview() {
+		widget = new Paragraph();
+		getPreview().addStyleName("form-control");
+		return widget;
+	}
+
+	@Override
+	protected Widget getWidget() {
+		return this.widget;
+	}
+
+	@Override
+	protected void setValidationTriggers() {
+
+	}
+
+	@Override
+	protected void setValueImpl(String value) {
+
```

Comment:
Please implement `setValueImpl()` so that generic logic can pre-fill form items as correct data is returned from the API. Probably in each extending class, as it will be handled differently.